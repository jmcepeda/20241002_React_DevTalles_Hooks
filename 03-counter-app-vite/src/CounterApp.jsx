import PropTypes from "prop-types"

import { useState } from "react";

// El Hook useState se utiliza para crear un estado local en un componente funcional
// Este es el Primer Hook
// Todos los nombres de los Hooks empiezan con "use"

// Siempre que sea posible es preferible que las funciones vayasn fuera del Funcional Component, Fuera del Componente Funcional, pues la función ocupara menos recursos pues solo ocupa espacio en memoria una sola vez, indepenidientemente del número de veces que se renderice el componente.
const handleAppOut = (event) => {
  console.log(event);
}

// Mi Primer Hook

export const CounterApp = ({value}) => {
  
  // Aqui lo que estoy haciendo es desectrucurar la salida de la funcion useState en un array
  const [counter_v0,setCounter_v0]=useState(0)
  const [counter_v1,setCounter_v1]=useState(1)
  const [counter_v2,setCounter_v2]=useState(10)
  const [counter_v3,setCounter_v3]=useState(100)

  
  const handleAppIn = () => {
    // console.log(event, newValue);
         
    console.log("value antes del Evento:", value)
    console.log("+1")
 
    value=11000
    console.log("value despues del evento en función handleAppIn: ", value)
    // value=1000; Este codigo no va a funcionar, no refleja el cambio del valor de la variable en el Componente
    // En React, Para que haya cambios en los valores que representa el Componente, una vez que se renderiza, hay que indicarle a React que vha habido ese cambio para que vuelva a Renderizar el componente, este React lo hace de una forma muy eficiente
  }

  const handleApp_v0= () => {
    setCounter_v0(() => {
      return 1000; 
      // Es muy interesante como funciona el useState, si el valor del useState no cambia, no se vuelve a renderizar el HTML hasta que no cambia el valor del useState, esto hace que REACT sea muy eficiente y rápdio
    })
  }


  const handleApp_v1 = () => {
    // console.log(event, newValue);
    // Esto no funciona porque se ha definido counter como una Constante 
    // Además, para que se actualice correctamete, hay que usar el Hooks, 
    // Si no recuerdo Mal, el Hook crea la variable de estado a actualizar y te da también la función para actualizar su visualViewport, para asegurarse de que se hace de forma ordenada
    // counter++;
    setCounter_v1(counter_v1 + 1); // Esto es la manera correcta de actualizar el estado en React con el Hook useState
    // value=1000; Este código no va a funcionar, no refleja el cambio del valor de la variable en el Componente
    // En React, Para que haya cambios en los valores que representa el Componente, una vez que se renderiza, hay que indicarle a React que vha habido ese cambio para que vuelva a Renderizar el componente, este React lo hace de una forma muy eficiente
    // 
    // Es muy interesante también, tener en cuenta que aunque tenga varios Hooks en una función, solo se renderiza una vez el HTML, solo cuando han terminado la ejecución de todos los Hooks de la Función se renderiza de nuevo el HTML. Lo que hace que sea muy eficiente y rápido
  }
  const handleApp_v2 = () => {

    setCounter_v2(()=>{
      return counter_v2*3 -1

    }); // Esto es la manera correcta de actualizar el estado en 
  }
  const handleApp_v3 = () => {

    setCounter_v3((new_counter_v3)=>{
      return new_counter_v3*5-5

    }); // Esto es la manera correcta de actualizar el estado en 
  }
  
  return (
    <>


      <h2>counter - useState_v0: { counter_v0 }</h2>
      <h1>CounterApp</h1>
      <h2>counter - useState_v1: { counter_v1 }</h2>
      <br />
      <h2>counter - useState_v2: { counter_v2 }</h2>
      <br />
      <h2>counter - useState_v3: { counter_v3 }</h2>
      <br />
      <h2>Variable Generica pasada por jsx: { value }</h2>
      {/* Esta sería la manera de definir correctamwnte la función en JavaScript,
      Pero los nuevos estandar de javascrit permite simplificar la definición 
      Cuando el argumento del evento coincide con el argumento que se pasa a la función, es por esto que directamente se puede poner la función en el evento y se puede simplificar, puesto que el argumento del evento y de la función vinculada coinciden  */}

      <button onClick={()=>{handleApp_v0()}}>1.000 Hook useState_v0</button>
      
      <button onClick={()=> {handleApp_v1()}}>+1 Hooks useState_v1</button>
      <br />

      <button onClick={()=> {handleApp_v2()}}>*3 -1 Hooks useState_v2</button>
      <br />

      <button onClick={()=> {handleApp_v3()}}>*5 -5 Hooks useState_v3</button>
      <br />

      <button onClick={(event)=>handleAppIn(event)}>
      +1</button>
     
      <button onClick={(event)=>handleAppIn(event,"Hola Mundo")
      }>+1</button>

      <br />

      <button onClick={handleAppOut}>+1</button> 
      
    </>
  )
}

CounterApp.propTypes ={
    value: PropTypes.number.isRequired, // Requerido y debe ser un número
};

CounterApp.defaultvalue ={
    value: 0, // Valor por defecto
}


