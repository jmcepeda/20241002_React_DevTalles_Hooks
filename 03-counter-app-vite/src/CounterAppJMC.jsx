import PropTypes from "prop-types"


import { useState } from "react"

export const CounterAppJMC = ({value=0}) => {

console.log("Render")
  
const [counter,setCounter]=useState(value)

// Es muy importante tener en cuenta que cada vez que cambia el estado vuelve a dispararae el componente y vuelve a renderizarse el componente

  const handleCounter= (id)=> {
    setCounter(()=>{
        let result=0
        if (id===0){
            result=counter+1
        } else if (id===1) {
            result=counter-1
        } else if (id===2) {
            result=value
        }
        return result

        // Es muy interesante como funciona el useState, si el valor del useState no cambia, no se vuelve a renderizar el HTML hasta que no cambia el valor del useState, esto hace que REACT sea muy eficiente y r√°pido
    })
  }

  return (
    <>
        <h1>Counter App JMC</h1>
        <h2>{ counter }</h2>
        <button onClick={()=>{handleCounter(0)}}>+1</button>
        <button onClick={()=>{handleCounter(1)}}>-1</button>
        <button aria-label="btn-reset" onClick={()=>{handleCounter(2)}}>Reset</button>

    </>
  )
}

CounterAppJMC.propTypes = {
    value: PropTypes.number.isRequired,
}

// CounterAppJMC.defaultProps = {
//     value: 0,
// }

