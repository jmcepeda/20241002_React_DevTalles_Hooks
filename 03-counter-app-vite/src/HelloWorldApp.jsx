// Es muy importante Tener en Cuenta que los componentes de React hay que designarlos usando Pascal <OpenCamelCase>
// Siempre la primera Letra de los componentes debe ser en may√∫scula

export function App() {
    // Document.createElement
    return <h1>Hola Mundo - Ahammm</h1>
}

 const App_two = () => {
    // Document.createElement
    return <h2>Hello World App!!</h2>
}

export default App_two;




