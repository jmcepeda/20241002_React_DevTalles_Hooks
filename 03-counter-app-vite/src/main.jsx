import React from 'react'
import ReactDOM from 'react-dom/client'
import App_two, {App} from './HelloWorldApp'
import FirstApp from './FirstApp'
import { CounterApp } from './CounterApp'

import './styles.css'
import { CounterAppJMC } from './CounterAppJMC'



// Create a React component

// Los Componenetes de React son funciones de </JavaScript>
// Es por esto que a React se le llama Funcional Component. Porque los COMPONENTES SON FUNCIONES

// Es muy importante Tener en Cuenta que los componentes de React hay que designarlos usando Pascal <OpenCamelCase>
// Siempre la primera Letra de los componentes debe ser en mayúscula


ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        {/* Para Mandar Prop como un número, hay que meter el valor entre corchetes, si se manda entre comillas el Prop se enviará como un String */}
        <CounterAppJMC value={100}/>
        <CounterApp value={55}/>
        <App />  {/* Aquí se renderiza el componente */}
        <App_two/>
        <FirstApp title='Props Title Ejemplo Rabito duro' ardecor="El Orto de tu Madre" subTitle="Esto es un Ejemplo de Props Props Subtitle" valor={345}/>
        <FirstApp title='Este es un Prop de Título de Segundo Comoponente FirstApp' />
        <FirstApp subTitle='Este es un Prop de SubTítulo de Segundo Comoponente FirstApp' />
        <FirstApp valor={10000} />
        <FirstApp valor={10000} />
        {/* Esta es la manera de pasar un número como un props USando llaves {} */}
    </React.StrictMode>

);

// Vite tiene la habilidad de cambiar módulos en caliente.App
// Esto no es propio de REact, es una funcionalidad de VITE
// Lo que hace que el funcionamiento sea muy rápido, creo que en Development