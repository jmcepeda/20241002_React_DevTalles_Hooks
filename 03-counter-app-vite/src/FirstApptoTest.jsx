
import React from 'react';
const nombre="TroncoMan"
const newObject={id: 22, name: "Alberto", age:23, Alias: "Al"};
const newArray=[1,2,3,4,5,6,7,8,9];
import {malditosBastardos} from './HelloWorldApp'
import PropTypes from 'prop-types'


const getBastardoById = (Id) => {
  return malditosBastardos.find((bastardo)=> {
    return bastardo.id === Id;
  }
  )
}

// No podemos usar funciones asincronas como argumentos en código HTML, pues devuelven objetos y React no gestiona Objetos

// const getSuma = async(Id) => {
//     return 2+2
// }

const getSuma = (a,b) => {
    return a + b
}

// Siempre que sea posible es mejor meter todas las funciones, importaciones creación de variables fuera del las Function Component 
// Es más eficiente y funciona todo mejor y más rápido

// PropTypes Tipos de las Properties 
// Esto viene instalado por defecto en VITE

//Es mejor definir los valores popr defecto con los DefaultProperties


const FirstApptoTest = ({title="No hay Título con DefaultProps", subTitle="No hay Subtítulo con defaultProps", valor=0.1 }) => {
  
  console.log(title);

  // Hacerlo así, es un problema, porque complica mi componente, 
  // lo hace dificil de leer 
  // ¿Como obligo a que se exija el que se rellenen los Props
  // Sin tener que complicar mi componente

  // if(!subTitle) {
  //   throw new Error("The SubTitle no Exist")
  // }
  

  return (
    <>
        <h1 data-testid="test-title">{title}</h1>
        <h2>{subTitle}</h2>
        <h3>{subTitle}</h3>
        {/* <h3>{subTitle}</h3> */}
        <h2>Este es el Valor Numerico con Operación con un Props: {valor*2.5}</h2>
        {/* <h1>Hola este es mi nombre y es una variable: {nombre}</h1> */}
        <h2>Resultado de la Operación 2+2 ={2+2}</h2>
        <h3>Propiedades de un Elemento: {newObject.name}</h3>
        <h4>Este es el resultado de un array. Coloca elementos Independientes: {newArray}</h4>
        {/* En caso de que tuviera que enviar un Objeto al DOM, podría hacerlo de la siguiente manera. Convierto en String el Objeto */}
        <div>
          <p>Propiedades de un Objeto en el DOM:</p>
          <code> {JSON.stringify(newObject)}</code>
        </div>
        <h2></h2>
        <ol><b>Este es el Bastardo que he encontrado</b></ol>
        <li>name: <b>{getBastardoById(2).name}</b></li>
        <li>owner: <b>{getBastardoById(2).owner}</b></li>
        <li>Este es el resultado de sumar dos Números: <b>{getSuma(4,6)}</b></li>
    </>
  )
}

export default FirstApptoTest


// Tanto los PropTypes como los defaultProps se colocan siempre al final del documento. Esto es lo que dice el Estandar ES6
FirstApptoTest.propTypes = {
  title: PropTypes.string.isRequired,
  subTitle: PropTypes.string.isRequired,
  valor: PropTypes.number,
}

// FirstApptoTest.defaultProps = {
//   title: 'No hay Título con DefaultProps',
//   subTitle: 'No hay Subtítulo con defaultProps',
//   valor: 0.1,

// };
