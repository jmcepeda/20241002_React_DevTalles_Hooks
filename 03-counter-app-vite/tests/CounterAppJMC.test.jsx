import { expect } from "vitest";
import { CounterAppJMC } from "../src/CounterAppJMC";
import { fireEvent, render,screen } from "@testing-library/react";


describe('Vamos a hacer las Pruebas del Componente CounterAppJMC', () => { 
    const value=100;
    test('Prueba de SnapShot ', () => {
      const {container}=render(<CounterAppJMC value={value}/>)
      expect(container).toMatchSnapshot();
    })

    test('Prueba de Valor Iniical de 100 ', () => {
        render(<CounterAppJMC value={value}/>);
        expect(screen.getByText('100')).toBeTruthy();
        expect(screen.getByText(100)).toBeTruthy();
        expect(screen.getAllByRole("heading", {level: 2})[0].innerHTML).toBe("100");
    })

    test('Debe incrementar en uno al pinchar en el Botón "+1" ', () => {
      render(<CounterAppJMC value={value}/>);
      const botonMas=screen.getByText('+1');
      fireEvent.click(botonMas);
      expect(screen.getAllByRole("heading", {level: 2})[0].innerHTML).toBe("101");
      
    })
    test('Debe Decrementar en uno al pinchar en el Botón "-1" ', () => {
      render(<CounterAppJMC value={value}/>);
      const botonMas=screen.getByText('-1');
      fireEvent.click(botonMas);
      expect(screen.getAllByRole("heading", {level: 2})[0].innerHTML).toBe("99");
      // screen.debug();
      
    })
    test('Debe Resetearse en 100 al pinchar en el Botón "Reset" ', () => {
      render(<CounterAppJMC value={value}/>);
      const botonMas =screen.getByText("+1");
      fireEvent.click(botonMas);
      fireEvent.click(botonMas);
      fireEvent.click(botonMas);
      // const botonReset=screen.getByText('Reset');
      const botonReset=screen.getByRole("button",{name:'btn-reset'});
      fireEvent.click(botonReset);
      // screen.debug();
      expect(screen.getAllByRole("heading", {level:2})[0].innerHTML).toBe("100");
      
      
    })


    
    
 })