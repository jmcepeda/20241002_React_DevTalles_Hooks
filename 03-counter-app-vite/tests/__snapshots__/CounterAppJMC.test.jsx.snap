// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vamos a hacer las Pruebas del Componente CounterAppJMC > Prueba de SnapShot  1`] = `
<div>
  <h1>
    Counter App JMC
  </h1>
  <h2>
    100
  </h2>
  <button>
    +1
  </button>
  <button>
    -1
  </button>
  <button
    aria-label="btn-reset"
  >
    Reset
  </button>
</div>
`;
